taichi/transforms/bit_loop_vectorize.cpp

```diff
@@ -38,105 +38,108 @@ class BitLoopVectorize : public IRVisitor {
-        // rewrite the previous GlobalPtrStmt's return type from *qit to
-        // *phy_type
-        auto ptr = stmt->src->cast<GlobalPtrStmt>();
-        auto ptr_physical_type = TypeFactory::get_instance().get_pointer_type(
-            quant_array_physical_type, false);
-        DataType new_ret_type(ptr_physical_type);
-        ptr->ret_type = new_ret_type;
-        ptr->is_bit_vectorized = true;
-        // check if j has offset
-        if (ptr->indices.size() == 2) {
-          auto diff = irpass::analysis::value_diff_loop_index(ptr->indices[1],
-                                                              loop_stmt, 1);
-          // TODO: temporarily we only support [j - 1] and [j + 1]
-          //       the general case should be easy to implement
-          if (diff.linear_related() && diff.certain() &&
-              (diff.low == 1 || diff.low == -1)) {
-            // construct ptr to x[i, j]
-            auto indices = ptr->indices;
-            indices[1] = loop_stmt->body->statements[1].get();
-            auto base_ptr =
-                std::make_unique<GlobalPtrStmt>(ptr->snode, indices);
-            base_ptr->ret_type = new_ret_type;
-            base_ptr->is_bit_vectorized = true;
-            // load x[i, j](base)
-            DataType load_data_type(quant_array_physical_type);
-            auto load_base = std::make_unique<GlobalLoadStmt>(base_ptr.get());
-            load_base->ret_type = load_data_type;
-            // load x[i, j + 1](offsetted)
-            // since we are doing vectorization, the actual data should be x[i,
-            // j + vectorization_width]
-            auto vectorization_width = data_type_bits(load_data_type);
-            auto offset_constant =
-                std::make_unique<ConstStmt>(TypedConstant(vectorization_width));
-            auto offset_index_opcode =
-                diff.low == -1 ? BinaryOpType::sub : BinaryOpType::add;
-            auto offset_index = std::make_unique<BinaryOpStmt>(
-                offset_index_opcode, indices[1], offset_constant.get());
-            indices[1] = offset_index.get();
-            auto offset_ptr =
-                std::make_unique<GlobalPtrStmt>(ptr->snode, indices);
-            offset_ptr->ret_type = new_ret_type;
-            offset_ptr->is_bit_vectorized = true;
-            auto load_offsetted =
-                std::make_unique<GlobalLoadStmt>(offset_ptr.get());
-            load_offsetted->ret_type = load_data_type;
-            // create bit shift and bit and operations
-            auto base_shift_offset =
-                std::make_unique<ConstStmt>(TypedConstant(load_data_type, 1));
-            auto base_shift_opcode =
-                diff.low == -1 ? BinaryOpType::bit_shl : BinaryOpType::bit_sar;
-            auto base_shift_op = std::make_unique<BinaryOpStmt>(
-                base_shift_opcode, load_base.get(), base_shift_offset.get());
+    if (auto ptr_type = stmt->src->ret_type->cast<PointerType>()) {
```

Comment:
Theoretically, the `src` of `LoadStmt` should have been of `PointerType` (basically you load from an address/pointer), otherwise it might be a bug. Ideally we should replace it with a `TI_ASSERT()`?